server/pulp/server/async/scheduler.py

```diff
@@ -186,57 +186,72 @@ def monitor_events(self):
         while True:
             time.sleep(self.FREQUENCY)
             try:
-                self.check_workers()
+                self.check_celery_processes()
             except Exception as e:
                 _logger.error(e)
 
-    def check_workers(self):
+    def check_celery_processes(self):
         """
-        Look for missing workers, and dispatch a cleanup task if one goes missing.
+        Look for missing Celery processes, log and cleanup as needed.
 
-        To find a missing worker, filter the Workers model for entries older than
+        To find a missing Celery process, filter the Workers model for entries older than
         utcnow() - WORKER_TIMEOUT_SECONDS. The heartbeat times are stored in native UTC, so this is
-        a comparable datetime.
-
-        For each missing worker found, dispatch a _delete_worker task requesting that the resource
-        manager delete the Worker and cleanup any associated work.
+        a comparable datetime. For each missing worker found, call _delete_worker() synchronously
+        for cleanup.
 
-        This method logs and the debug and error levels.
+        This method also checks that at least one resource_manager and one scheduler process is
+        present. If there are zero of either, log at the error level that Pulp will not operate
+        correctly.
         """
-        msg = _('Looking for workers missing for more than %s seconds') % self.WORKER_TIMEOUT_SECONDS
+        msg = _('Checking if pulp_workers, pulp_celerybeat, or pulp_resource_manager '
+                'processes are missing for more than %s seconds') % self.CELERY_TIMEOUT_SECONDS
         _logger.debug(msg)
-        oldest_heartbeat_time = datetime.utcnow() - timedelta(seconds=self.WORKER_TIMEOUT_SECONDS)
-        worker_criteria = Criteria(filters={'last_heartbeat': {'$lt': oldest_heartbeat_time}},
-                                   fields=('_id', 'last_heartbeat'))
-        worker_list = list(resources.filter_workers(worker_criteria))
+        oldest_heartbeat_time = datetime.utcnow() - timedelta(seconds=self.CELERY_TIMEOUT_SECONDS)
+        all_worker_criteria = Criteria(filters={}, fields=('_id', 'last_heartbeat'))
+        worker_list = list(resources.filter_workers(all_worker_criteria))
+        resource_manager_count = 0
+        scheduler_count = 0
         for worker in worker_list:
-            msg = _("Workers '%s' has gone missing, removing from list of workers") % worker.name
+            if worker['last_heartbeat'] < oldest_heartbeat_time:
+                msg = _("Worker '%s' has gone missing, removing from list of workers") % worker.name
+                _logger.error(msg)
+                _delete_worker(worker.name)
+            elif worker['name'].startswith(SCHEDULER_WORKER_NAME):
+                scheduler_count = scheduler_count + 1
+            elif worker['name'].startswith(RESOURCE_MANAGER_WORKER_NAME):
+                resource_manager_count = resource_manager_count + 1
```

Comment:
It may be worthwhile to add a debug-level logging statement here for total number of schedulers and resource managers.